### ISSUE TITLE
Ability to bypass, or set cache key per URL

### ISSUE THREAD

**Issue by nickdunn** [putyourlightson/craft-blitz#302]
I have quite a niche requirement, so stick with me :-)

The site I'm developing has several "audience types", let's call them Individuals and Professionals. There is an audience switcher at the top of the site listing these audience types, allowing you to toggle between them:

- Individuals `/individuals`
- Professionals `/professionals`

The vast majority of pages sit under these audience sections. However there are some global sections e.g. `/about-us` or `/contact` that don't sit within an audience type, but we still want to keep the audience type selected in the page. For these global sections we are currently going down the route of setting a cookie each time you switch audience types, to track the last one you actively selected, so that we can select the right tab at the top of a global page. Otherwise we wouldn't know which one to select and we'd have to default to none (or "Individuals" as the first one) — which is what we do if no cookie is present.

The entire primary navigation changes based on the audience type; so the issue is we can no longer cache the entire page because we need to change the markup based on a cookie value. 

We don't want to switch the navigation using JavaScript because we have 5 possible audience types, which means loading 5 entire navigations into the page, meaning potentially hundreds of database queries, and potentially a janky switcheroo when JavaScript kicks in. We want to keep the pages cached server-side.

In the past I've configured Varnish to use the Vary header (https://www.fastly.com/blog/best-practices-using-vary-header) based on cookies, so that specific cookie values can be used to either bypass, or serve different cached versions of pages.

Would this be possible using Blitz? I was thinking maybe some way of overriding the cache key that Blitz uses for each page, a compound key combining the page URL with other values, such as value of a cookie. 

```
{% do craft.blitz.options({
    cachingEnabled: true,
    cacheKey: craft.request.getUrl() ~ craft.request.getCookie('audienceType')
}) %}
```

Am I barking up the wrong tree entirely? Is there another strategy I haven't considered?

**Comment by bencroker** [comment]
Blitz provides [dynamic content](https://putyourlightson.com/plugins/blitz#dynamic-content) tags for this purpose, although I'm not sure how well it will work for a main navigation as it requires an AJAX request so may end up being janky. Blitz statically caches full pages so there's no obvious workaround besides changing your URL structure or disabling caching of those pages. You could possibly get away with using query string parameters, if that would be an option for you, for example `/about-us?type=individuals`.

**Comment by nickdunn** [comment]
Thanks Ben. Yeah we’ve looked at the dynamic AJAX requests and this feels a little too risky.

I guess where I was going was the ability to change the cache key that Blitz uses. So we could override the siteUri that’s currently used to “trick” Blitz into serving something from a cache still...

**Comment by bencroker** [comment]
> I guess where I was going was the ability to change the cache key that Blitz uses. So we could override the siteUri that’s currently used to “trick” Blitz into serving something from a cache still...

That's not currently possible. While I can see the use-case, your example shows how to add cache keys but not how to retrieve them. Requiring the execution of PHP to handle this would mean sacrificing a lot of the performance gains that Blitz is specifically trying to afford you.

**Comment by nickdunn** [comment]
Makes sense, thanks for explaining.

How does the Twig-based config work? Does Blitz pick this up the first time a page is cached only?

**Comment by bencroker** [comment]
> How does the Twig-based config work? Does Blitz pick this up the first time a page is cached only?

Yes, exactly.

**Comment by nickdunn** [comment]
Thinking aloud here, in case anyone else has this same use case...

I can capture the value of the audience type cookie using htaccess

```
RewriteCond %{HTTP_COOKIE} audienceType=([^;]+)
```

And then pass it on as a querystring parameter to every request:

```
RewriteRule (.+) index.php?p=$1&audienceType=%1 [QSA,L]
```

I can then configure Blitz to vary based on querystring (`'queryStringCaching' => 1`).

I _think_ this solves my use case to retain Blitz caching, but serve up different responses for a single URL based on the value of a cookie.

**Comment by bencroker** [comment]
Thanks for sharing that! It may solve your use-case, but note that this solution is specific to using the file cache storage driver, server redirects and Apache.

**Comment by goraxan** [comment]
Thanks @nickdunn so much for pointing me on the right direction, I've managed to do the same in nginx.
This is how I did it in case anyone is facing the same problem. I've modified the `QUERY_STRING` cgi param to include a "fake" query parameter based on a cookie value:
From the standard:
`fastcgi_param QUERY_STRING       $query_string`
To:
```
set $querystr $query_string;
if ($cookie_user_initials) {
    set $querystr $query_string&useini=$cookie_user_initials;
}

fastcgi_param QUERY_STRING    $querystr;
```
In my case the page header has to show the user initial in the menu, so I'm creating a cookie called user_initials with just the name's first letter when the user logs in and then caching different versions of the same page depending on the cookie value.
